testnet/stacks-node/src/tests/nakamoto_integrations.rs

```diff
@@ -3455,3 +3497,498 @@ fn forked_tenure_is_ignored() {
+            .unwrap()
+            .text()
+            .unwrap();
+        let expected_result = format!("stacks_node_stacks_tip_height {block_height_pre_3_0}");
+        assert!(res.contains(&expected_result));
+    }
+
+    info!("Nakamoto miner started...");
+    blind_signer(&naka_conf, &signers, proposals_submitted.clone());
+
+    let socket = naka_conf
+        .node
+        .rpc_bind
+        .to_socket_addrs()
+        .unwrap()
+        .next()
+        .unwrap();
+    let signer_client = stacks_signer::client::StacksClient::new(
+        StacksPrivateKey::from_seed(&[0, 1, 2, 3]),
+        socket,
+        naka_conf
+            .connection_options
+            .block_proposal_token
+            .clone()
+            .unwrap_or("".into()),
+        false,
+    );
+
+    // there hasn't been a successful nakamoto sortition yet, so expect an error
+    assert!(
+        matches!(
+            sortitions_view.refresh_view(&signer_client).unwrap_err(),
+            ClientError::UnexpectedSortitionInfo
+        ),
+        "Sortitions view should fail to refresh if there are no successful nakamoto sortitions yet",
+    );
+
+    // first block wakes up the run loop, wait until a key registration has been submitted.
+    next_block_and(&mut btc_regtest_controller, 60, || {
+        let vrf_count = vrfs_submitted.load(Ordering::SeqCst);
+        Ok(vrf_count >= 1)
+    })
+    .unwrap();
+
+    // second block should confirm the VRF register, wait until a block commit is submitted
+    next_block_and(&mut btc_regtest_controller, 60, || {
+        let commits_count = commits_submitted.load(Ordering::SeqCst);
+        Ok(commits_count >= 1)
+    })
+    .unwrap();
+
+    let mut signer_db =
+        SignerDb::new(format!("{}/signer_db_path", naka_conf.node.working_dir)).unwrap();
+
+    // Mine some nakamoto tenures
+    //  track the last tenure's first block and subsequent blocks so we can
+    //  check that they get rejected by the sortitions_view
+    let mut last_tenures_proposals: Option<(StacksPublicKey, NakamotoBlock, Vec<NakamotoBlock>)> =
+        None;
+    // hold the first and last blocks of the first tenure. we'll use this to submit reorging proposals
+    let mut first_tenure_blocks: Option<Vec<NakamotoBlock>> = None;
+    for i in 0..5 {
+        next_block_and_mine_commit(
+            &mut btc_regtest_controller,
+            60,
+            &coord_channel,
+            &commits_submitted,
+        )
+        .unwrap();
+
+        sortitions_view.fresh = false;
+        sortitions_view.refresh_view(&signer_client).unwrap();
+
+        // check the prior tenure's proposals again, confirming that the sortitions_view
+        //  will reject them.
+        if let Some((ref miner_pk, ref prior_tenure_first, ref prior_tenure_interims)) =
+            last_tenures_proposals
+        {
+            let valid = sortitions_view
+                .check_proposal(&signer_client, &signer_db, prior_tenure_first, miner_pk)
```

Comment:
I love all these tests, but is it possible to include the check_proposal tests as unit tests in signer directly by simulating data in the database rather than relying on the miner interactions? Or is it too hard to build the blocks manually? I know I was having a hard time doing this in my own PR so ignore me if it is too difficult.

Reply:
Yep, will add